### ISSUE TITLE
Supporting legacy versions of CABLE

### ISSUE THREAD

**Issue by SeanBryan51** [CABLE-LSM/benchcab#63]
Currently `benchcab` only supports the CABLE v3 build system. This is fine for the use case where a developer is attempting merge a branch into the trunk because the development branch will be forced to adopt the newer CABLE v3 build system.

**Problem**: there is also the use case where a developer wants to do a scientific evaluation for older versions of CABLE which use a legacy build system.

For example, when developers are preparing to merge some old branch of CABLE (label as `branch_old`) into the trunk, it may be easier to first merge an older revision of the trunk into `branch_old` as opposed to merging the head of the trunk into `branch_old` (as the former will likely be easier to resolve any merge conflicts). The developer may then want to compare the output between a new version of `branch_old` with the original `branch_old` to verify if any changes in the new version have impacted the results.

This would require `benchcab` to be able to build and run legacy versions of CABLE.

Possible solutions:
1. Write separate "drivers" that build CABLE for each legacy version into `benchcab`.
    - **pro:** this might not take much work as there is only a finite number of legacy build systems for CABLE
    - **con:** requires updating `benchcab` every time we encounter a build system that is not supported
2. Allow the user to specify the commands/instructions required to build a particular branch in the config file. These commands are then run instead of the default CABLE v3 build scripts.
    - **pro:** `benchcab` can support any arbitrary build system
    - **con:** requires extra work for the user
    - **con:** it is difficult to make this an automated step in the work flow as the user must know the directory structure of the working directory
3. Update the legacy build system to use the CABLE v3 build system.
    - **con:** this requires major refactoring of the code base for legacy branches of CABLE which should be avoided
4. Allow the user to specify the path to a prebuilt CABLE executable.
    - **pro:** this is already an issue #33 
    - **pro:** this functionality is great for this specific use case. However, **con:** this is not ideal for the use case where a developer is testing a branch that is about to be merged into the trunk as this would allow for potential uncommitted changes to leak into the test suite.

**Comment by SeanBryan51** [comment]
@ccarouge I think I'm leaning towards option 4. What do you think?

@rkutteh I am interested if you have any thoughts on this as well

**Comment by ccarouge** [comment]
I think option 2 can be very easy to implement. Probably faster than option 4. And this:

>con: it is difficult to make this an automated step in the work flow as the user must know the directory structure of the working directory

can be avoided. If we ask the user to give a path to a build script *relative* to the root of their source code directory. It's easy for the user to know, they don't need to know where it will live when `benchcab` runs, `benchcab` can find it as it knows where the root of each source directory is.

It also makes it easier for people to test other users' branches without having to do the checkout and compilation by hand first.

Probably, we will want both options 2 and 4 in eventually. It's just a question of which one to put in first.

**Comment by ccarouge** [comment]
Actually, option 4 is probably as easy to implement as option 2. So @rkutteh , which one would work best for you:

- specify an executable existing on the filesystem
- specify the relative path to the build script in the branch

**Comment by rkutteh** [comment]
- The most elegant and "black box" solution in my view is (1), but of course it requires more work than the others, although we only need to worry about v2 and v3 for now (and ensure the build workflow is not needlessly altered in future CABLE versions). 
-  My second choice would be (2). 
- On the other hand (3) is a definite no for me. 
- Lastly, I don't like (4) at all, partly for the 'con' reason mentioned, but in general whatever is getting tested should be committed already as this would allow others to reproduce the test. In addition, both builds being compared should be compiled in the same manner (same compiler flags) for a true comparison of results. There are simply too many things in this scenario that could corrupt an accurate comparison of results (e.g., we don't know how the prebuilt exec was compiled, what exactly was compiled at the time,...). In my view, benchcab should only be allowed to compare versions already in a repo as this keeps things "in check" so to speak.

**Comment by rkutteh** [comment]
@ccarouge @SeanBryan51 as per my comments above, I like (2) but not (4). In other words I prefer to "specify the relative path to the build script in the branch"

**Comment by ccarouge** [comment]
One of the use cases for option 4 is during development. When people develop, it would be good if they could quickly fire up a test on one site as a sanity check they haven't broken anything by inadvertence. Obviously, these test results would not be admissible to show the quality of a new feature. The problem is to quickly identify which tests have been done using a pre-built executable so no mistake is done. 

I'll see about option 1. I also like option 2 because if someone comes with yet a different build script, they can use `benchcab` before "we" (preferably "they") develop a new driver.

**Comment by rkutteh** [comment]
So again the problem with (4) is that even during development I may do a test today with my current working copy but forget after a bit more development what my working copy looked like when I did that test. Because I did not checkout the copy for the test from a repo then I do not really have a record for what the test corresponded to. Again to me this feature encourages poor development practices, but I guess for really incremental work it is useful to have.
